apiserver/apiserver_test.go

```diff
@@ -0,0 +1,327 @@
+package apiserver
```

Comment:
Many of the tests in this file would be good candidates for table driven tests (see https://github.com/golang/go/wiki/TableDrivenTests). For example most (all?) of the GET/UPDATE/LIST tests can be combined into a single test with a table defining the input/output parameters. Adding a new test only requires adding a new entry to the table (see snippet below). This also avoids the need for the extra helper functions at the bottom of the file.


```golang
func TestConfig(t *testing.T) {
    cases := []struct {
        name       string
        url        string
        method     string
        data       []byte
        wantStatus int
        wantErr    error
    }{
        {
            name:       "TestNotFoundGetConfig",
            url:        "/test/config/route-rule/namespace/name",
            method:     "GET",
            wantStatus: http.StatusNotFound,
            wantBody:   errInvalidType.Error(),
        },
        {
            name:       "TestInvalidConfigTypeGetConfig",
            url:        "/test/config/not-a-route-rule/namespace/name",
            method:     "GET",
            wantStatus: http.StatusBadRequest,
            wantBody:   errInvalidType.Error(),
        },
    }

    for _, c := range cases {
        api := makeAPIServer(mock.MakeRegistry())
        gotStatus, gotBody, err := makeAPIRequest(api, c.method, c.url, c.data)
        if err != nil {
            t.Errorf("%s: makeAPIRequest failed: %v", c.name, err)
        }
        if gotStatus != c.wantStatus {
            t.Errorf("%s: got status code %v, want %v", c.name, gotStatus, c.wantStatus)
        }
        if gotBody != c.wantBody {
            t.Errorf("%s: got body %q, want %q", c.name, gotBody, c.wantBody)
        }
    }
}
```